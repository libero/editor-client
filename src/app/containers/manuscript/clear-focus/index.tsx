import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import * as manuscriptEditorAction from 'app/actions/manuscript-editor.actions';

export const ClearFocus: React.FC<{ className?: string }> = ({ children, className }) => {
  const dispatch = useDispatch();
  const handleClick = useCallback(() => {
    dispatch(manuscriptEditorAction.removeFocusAction());
  }, [dispatch]);

  return (
    <div className={className} onClick={handleClick}>
      {children}
    </div>
  );
};
