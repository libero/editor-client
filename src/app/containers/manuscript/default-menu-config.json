{
  "button-group": [
    "save",
    "undo",
    "redo",
    "pdf"
  ],
  "menus": [
    "headings",
    "typography",
    "insert"
  ],
  "custom-buttons": []
}