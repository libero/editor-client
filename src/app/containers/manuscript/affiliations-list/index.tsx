import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';
import { useAffiliationStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedAffiliationFormDialog } from 'app/containers/affiliation-form-dialog';
import { ComponentWithId } from 'app/types/utility.types';

export const AffiliationsList: React.FC<ComponentWithId> = ({ id }) => {
  const classes = useAffiliationStyles();
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const editAffiliation = useCallback(
    (aff: Affiliation) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedAffiliationFormDialog,
          props: { affiliation: aff },
          title: 'Edit Affiliation'
        })
      );
    },
    [dispatch]
  );

  const addAffiliation = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedAffiliationFormDialog,
        props: {},
        title: 'Add Affiliation'
      })
    );
  }, [dispatch]);

  const renderAffiliation = (aff: Affiliation): React.ReactNode => (
    <li key={aff.id} className={classes.listItem}>
      <div className={classes.orderLabel}>{affiliations.length > 1 ? `(${aff.label})` : undefined}</div>
      <div className={classes.affiliationInfo}>{aff.getDisplayName()}</div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAffiliation.bind(null, aff)}>
        <EditIcon fontSize="small" />
      </IconButton>
    </li>
  );

  return (
    <section>
      <SectionContainer label="Affiliations" id={id}>
        <ul className={classes.root}>{affiliations.map(renderAffiliation)}</ul>
      </SectionContainer>
      <ActionButton
        title="Add Affiliation"
        variant="addEntity"
        onClick={addAffiliation}
        className={classes.addButton}
      />
    </section>
  );
};
