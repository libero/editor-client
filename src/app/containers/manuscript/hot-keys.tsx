import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { isMacOs } from 'react-device-detect';
import { GlobalHotKeys, configure } from 'react-hotkeys';

import { canApplyMarkToSelection, canRedoChanges, canUndoChanges } from '../../selectors/manuscript-editor.selectors';
import * as manuscriptActions from '../../actions/manuscript.actions';

configure({
  ignoreEventsCondition(event) {
    return false;
  }
});

const withMod = (shortcut: string): string => {
  const mod = isMacOs ? 'cmd' : 'ctrl';
  return `${mod}+${shortcut}`;
};

export const HotKeyBindings: React.FC<{}> = () => {
  const dispatch = useDispatch();

  const canUndo = useSelector(canUndoChanges);
  const canRedo = useSelector(canRedoChanges);
  const canApply = useSelector(canApplyMarkToSelection);

  const invokeUndo = useCallback(() => {
    if (canUndo) {
      dispatch(manuscriptActions.undoAction());
    }
  }, [dispatch, canUndo]);

  const invokeRedo = useCallback(() => {
    if (canRedo) {
      dispatch(manuscriptActions.redoAction());
    }
  }, [dispatch, canRedo]);

  const invokeToggleMark = useCallback(
    (mark: string) => () => {
      if (canApply(mark)) {
        dispatch(manuscriptActions.toggleMarkAction(mark));
      }
    },
    [dispatch, canApply]
  );

  const keymap = {
    UNDO: withMod('z'),
    REDO: [withMod('y'), withMod('shift+z')],
    BOLD: withMod('b'),
    ITALIC: withMod('i'),
    UNDERLINE: withMod('u')
  };

  const handlers = {
    UNDO: invokeUndo,
    REDO: invokeRedo,
    BOLD: invokeToggleMark('bold'),
    ITALIC: invokeToggleMark('italic'),
    UNDERLINE: invokeToggleMark('underline')
  };

  return <GlobalHotKeys allowChanges={true} keyMap={keymap} handlers={handlers} />;
};
