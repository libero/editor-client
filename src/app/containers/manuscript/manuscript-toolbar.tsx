import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IconButton } from '@material-ui/core';
import UndoIcon from '@material-ui/icons/Undo';
import RedoIcon from '@material-ui/icons/Redo';

import * as manuscriptActions from '../../actions/manuscript.actions';
import { canRedoChanges, canUndoChanges } from '../../selectors/manuscript.selectors';
import './styles.scss';

export const ManuscriptToolbar: React.FC = () => {
  const canUndo = useSelector(canUndoChanges);
  const canRedo = useSelector(canRedoChanges);

  const dispatch = useDispatch();

  const invokeUndo = useCallback(() => dispatch(manuscriptActions.undoAction()), [dispatch]);
  const invokeRedo = useCallback(() => dispatch(manuscriptActions.redoAction()), [dispatch]);

  const renderContent = () => (
    <div>
      <IconButton edge='start' disabled={!canUndo} onClick={invokeUndo}>
        {' '}
        <UndoIcon />{' '}
      </IconButton>
      <IconButton edge='start' disabled={!canRedo} onClick={invokeRedo}>
        {' '}
        <RedoIcon />{' '}
      </IconButton>
    </div>
  );

  return renderContent();
};
