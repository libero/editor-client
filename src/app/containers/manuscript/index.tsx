import React from 'react';
import { useSelector } from 'react-redux';
import { Backdrop, CircularProgress } from '@material-ui/core';

import { isManuscriptLoaded } from '../../selectors/manuscript.selectors';
import './styles.scss';
import { ManuscriptToolbar } from './manuscript-toolbar';
import { ManuscriptEditor } from './manuscript-editor';

const renderBackdrop = (): JSX.Element => (
  <Backdrop open={true}>
    <CircularProgress color="inherit" />
  </Backdrop>
);

export const ManuscriptContainer: React.FC = () => {
  const isLoaded = useSelector(isManuscriptLoaded);
  const renderContent = (): JSX.Element => (
    <React.Fragment>
      <ManuscriptToolbar />
      <ManuscriptEditor />
    </React.Fragment>
  );

  return isLoaded ? renderContent() : renderBackdrop();
};
