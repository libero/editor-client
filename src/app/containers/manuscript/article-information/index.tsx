import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';
import { useSelector, useDispatch } from 'react-redux';

import { SectionContainer } from 'app/components/section-container';
import { getArticleInformation } from 'app/selectors/manuscript.selectors';
import { useArticleInformationStyles } from 'app/containers/manuscript/article-information/styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ArticleInfoFormDialog } from 'app/containers/article-info-form-dialog';

export const ArticleInformation: React.FC<{}> = () => {
  const articleInfo = useSelector(getArticleInformation);
  const dispatch = useDispatch();
  const editArticleInfo = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ArticleInfoFormDialog,
        title: 'Article Information'
      })
    );
  }, [dispatch]);

  const classes = useArticleInformationStyles();
  return (
    <SectionContainer label="Article information" className={classes.sectionContainer}>
      <section className={classes.infoSection}>
        <div>
          <strong> Article DOI: </strong> {articleInfo.articleDOI}
        </div>
        <div>
          <strong> Publisher ID: </strong> {articleInfo.publisherId}
        </div>
      </section>
      <IconButton classes={{ root: classes.editButton }} onClick={editArticleInfo}>
        <EditIcon fontSize="small" />
      </IconButton>
    </SectionContainer>
  );
};
