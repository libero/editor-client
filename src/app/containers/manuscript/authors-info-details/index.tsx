import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import EditIcon from '@material-ui/icons/Edit';
import { isBoolean, isEqual } from 'lodash';
import Interweave from 'interweave';

import { SectionContainer } from 'app/components/section-container';
import { getAuthorAffiliations, getAuthors } from 'app/selectors/manuscript.selectors';
import { Person } from 'app/models/person';
import {
  useAuthorDetailStyles,
  useAuthorsDetailsListStyles
} from 'app/containers/manuscript/authors-info-details/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { stringifyEditorState } from 'app/utils/view.utils';
import { OrcidIcon } from 'app/assets/icons';
import { ComponentWithId } from 'app/types/utility.types';

export const AuthorsInfoDetails: React.FC<ComponentWithId> = ({ id }) => {
  const authors = useSelector(getAuthors);
  const classes = useAuthorsDetailsListStyles();
  const dispatch = useDispatch();

  const addAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  return (
    <section>
      <SectionContainer label="Author Information" id={id}>
        {authors.map((author) => (
          <AuthorInfo key={author.id} author={author} />
        ))}
      </SectionContainer>
      <ActionButton variant="addEntity" title="Add Author" onClick={addAuthor} className={classes.addAuthorButton} />
    </section>
  );
};

interface AuthorInfoProps {
  author: Person;
}

const AuthorInfo: React.FC<AuthorInfoProps> = React.memo(({ author }) => {
  const classes = useAuthorDetailStyles();
  const dispatch = useDispatch();
  const affiliations = useSelector(getAuthorAffiliations)(author);

  const editAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author },
        title: 'Edit Author'
      })
    );
  }, [author, dispatch]);

  return (
    <section className={classes.root}>
      <div className={classes.authorInfoContainer}>
        <div className={classes.authorInfoLine}>
          <strong> {author.getDisplayName()} </strong>
        </div>
        <div className={classes.authorInfoLine}>
          {author.isCorrespondingAuthor ? 'Corresponding Author: ' : undefined}
          {author.email}
        </div>
        <div className={classes.authorInfoLine}>
          {affiliations.map((aff) => (
            <div key={aff.id}>{aff.getDisplayName()}</div>
          ))}
        </div>
        <div className={classes.authorInfoLine}>
          {author.isAuthenticated ? <OrcidIcon className={classes.orcidIcon} /> : undefined}{' '}
          <a href={`https://orcid.org/${author.orcid}`} target="_blank" rel="noopener noreferrer">
            {author.orcid}
          </a>
        </div>
        {isBoolean(author.hasCompetingInterest) ? (
          <div className={classes.authorInfoLine}>
            {author.hasCompetingInterest
              ? `Competing interest: ${author.competingInterestStatement}`
              : 'No Competing interest declared'}
          </div>
        ) : undefined}
        {author.bio ? (
          <div className={classes.authorBio}>
            <Interweave content={stringifyEditorState(author.bio)} />
          </div>
        ) : undefined}
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAuthor}>
        <EditIcon fontSize="small" />
      </IconButton>
    </section>
  );
}, isEqual);
