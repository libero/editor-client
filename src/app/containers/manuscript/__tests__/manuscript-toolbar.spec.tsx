import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { IconButton, AppBar, Toolbar } from '@material-ui/core';
import { mount } from 'enzyme';

import { getLoadableStateSuccess } from '../../../utils/state.utils';
import { ManuscriptToolbar } from '../manuscript-toolbar';
import * as manuscriptActions from '../../../actions/manuscript.actions';

jest.mock('@material-ui/core');

describe('<ManuscriptToolbar />', () => {
  const mockStore = configureMockStore([]);

  beforeEach(() => {
    (AppBar['render'] as jest.Mock).mockImplementationOnce((props) => <>{props.children}</>);
    (Toolbar['render'] as jest.Mock).mockImplementationOnce((props) => <>{props.children}</>);
    (IconButton['render'] as jest.Mock).mockImplementationOnce((props) => (
      <div data-cmp='icon-button'>{props.children}</div>
    ));
  });

  afterEach(() => {
    (AppBar['render'] as jest.Mock).mockReset();
    (Toolbar['render'] as jest.Mock).mockReset();
    (IconButton['render'] as jest.Mock).mockReset();
  });

  it('dispatches an event on undoClick', () => {
    const store = mockStore({
      manuscript: getLoadableStateSuccess({
        past: [{}],
        present: {},
        future: [{}]
      })
    });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ManuscriptToolbar />
      </Provider>
    );

    const undoBtnProps = wrapper.find(IconButton).at(1).props();

    expect(undoBtnProps.disabled).toBeFalsy();
    //call undo
    undoBtnProps.onClick(null);

    expect(store.dispatch).toBeCalledWith(manuscriptActions.undoAction());
  });

  it('dispatches an event on redoClick', () => {
    const store = mockStore({
      manuscript: getLoadableStateSuccess({
        past: [{}],
        present: {
          keywords: {}
        },
        future: [{}]
      })
    });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ManuscriptToolbar />
      </Provider>
    );

    const redoBtnProps = wrapper.find(IconButton).at(2).props();

    expect(redoBtnProps.disabled).toBeFalsy();
    //call redo
    redoBtnProps.onClick(null);

    expect(store.dispatch).toBeCalledWith(manuscriptActions.redoAction());
  });
});
