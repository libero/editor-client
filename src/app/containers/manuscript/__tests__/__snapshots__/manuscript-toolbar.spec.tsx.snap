// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManuscriptToolbar renders toolbar 1`] = `
<div
  data-cmp="AppBar"
>
  <div
    data-cmp="Toolbar"
  >
    <div
      data-cmp="IconButton"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
        />
      </svg>
    </div>
    <div
      data-cmp="ToggleButtonGroup"
    >
      <div
        data-cmp="ToggleButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 12v7H5v-7H3v7c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zm-6 .67l2.59-2.58L17 11.5l-5 5-5-5 1.41-1.41L11 12.67V3h2z"
          />
        </svg>
      </div>
      <div
        data-cmp="ToggleButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
          />
        </svg>
      </div>
      <div
        data-cmp="ToggleButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
          />
        </svg>
      </div>
      <div
        data-cmp="ToggleButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 2H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-8.5 7.5c0 .83-.67 1.5-1.5 1.5H9v2H7.5V7H10c.83 0 1.5.67 1.5 1.5v1zm5 2c0 .83-.67 1.5-1.5 1.5h-2.5V7H15c.83 0 1.5.67 1.5 1.5v3zm4-3H19v1h1.5V11H19v2h-1.5V7h3v1.5zM9 9.5h1v-1H9v1zM4 6H2v14c0 1.1.9 2 2 2h14v-2H4V6zm10 5.5h1v-3h-1v3z"
          />
        </svg>
      </div>
    </div>
    <div>
      <div
        data-cmp="Button"
      >
        PARAGRAPH
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <div
        data-cmp="Popper"
      >
        <div
          data-cmp="Paper"
        >
          <div
            data-cmp="ClickAwayListener"
          >
            <div
              data-cmp="MenuList"
            >
              <div
                data-cmp="MenuItem"
              >
                Heading 1
              </div>
              <div
                data-cmp="MenuItem"
              >
                Heading 2
              </div>
              <div
                data-cmp="MenuItem"
              >
                Heading 3
              </div>
              <div
                data-cmp="MenuItem"
              >
                Heading 4
              </div>
              <div
                data-cmp="MenuItem"
              >
                Paragraph
              </div>
              <div
                data-cmp="MenuItem"
              >
                Bulleted List
              </div>
              <div
                data-cmp="MenuItem"
              >
                Numbered List
              </div>
              <div
                data-cmp="MenuItem"
              >
                Preformat
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        data-cmp="Button"
      >
        FORMAT
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <div
        data-cmp="Popper"
      >
        <div
          data-cmp="Paper"
        >
          <div
            data-cmp="ClickAwayListener"
          >
            <div
              data-cmp="MenuList"
            >
              <div
                data-cmp="MenuItem"
              >
                Bold
              </div>
              <div
                data-cmp="MenuItem"
              >
                Italics
              </div>
              <div
                data-cmp="MenuItem"
              >
                Subscript
              </div>
              <div
                data-cmp="MenuItem"
              >
                Superscript
              </div>
              <div
                data-cmp="MenuItem"
              >
                Monospace
              </div>
              <div
                data-cmp="MenuItem"
              >
                Small Caps
              </div>
              <div
                data-cmp="MenuItem"
              >
                Underline
              </div>
              <div
                data-cmp="MenuItem"
              >
                Overline
              </div>
              <div
                data-cmp="MenuItem"
              >
                Strike Through
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        data-cmp="Button"
      >
        INSERT
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <div
        data-cmp="Popper"
      >
        <div
          data-cmp="Paper"
        >
          <div
            data-cmp="ClickAwayListener"
          >
            <div
              data-cmp="MenuList"
            >
              <div
                data-cmp="MenuItem"
              >
                Figure
              </div>
              <div
                data-cmp="MenuItem"
              >
                Inline Graphic
              </div>
              <div
                data-cmp="MenuItem"
              >
                Table
              </div>
              <div
                data-cmp="MenuItem"
              >
                Block Quote
              </div>
              <div
                data-cmp="MenuItem"
              >
                Box
              </div>
              <div
                data-cmp="MenuItem"
              >
                Equation
              </div>
              <div
                data-cmp="MenuItem"
              >
                File
              </div>
              <div
                data-cmp="MenuItem"
              >
                Inline Math
              </div>
              <div
                data-cmp="Divider"
              />
              <div
                data-cmp="MenuItem"
              >
                Reference Citation
              </div>
              <div
                data-cmp="MenuItem"
              >
                Figure Citation
              </div>
              <div
                data-cmp="MenuItem"
              >
                Table citation
              </div>
              <div
                data-cmp="MenuItem"
              >
                Footnote citation
              </div>
              <div
                data-cmp="MenuItem"
              >
                Equation citation
              </div>
              <div
                data-cmp="MenuItem"
              >
                File Citation
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="makeStyles-spacer-5"
    />
    <div
      className="makeStyles-toolbarMessage-6 error"
    >
      System offline
    </div>
  </div>
</div>
`;
